mypy/checker.py

```diff
@@ -1480,7 +1480,7 @@ def count_nested_types(self, typ: Instance, check_type: str) -> int:
     def visit_yield_stmt(self, s: YieldStmt) -> Type:
         return_type = self.return_types[-1]
         if isinstance(return_type, Instance):
-            if return_type.type.fullname() != 'typing.Iterator':
+            if not is_subtype(return_type, self.named_type('typing.Iterable')):
```

Comment:
Type checking `yield from` seem broken as well :-/

Actually, the right thing seems to be something like this: 

* Evaluate the `yield` expression in a context derived from the return type, resulting in the item type `T`.
* Construct a type `G = Generator[T, Any, Any]` (`T` is the item type).
* Verify that `G` is a subtype of the declared return type (i.e., reverse the order of the subtype check).

This would allow the return type to be `object`, for example, which would be correct (though not very useful).